experiment/image-bumper/main_test.go

```diff
@@ -0,0 +1,154 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import "testing"
+
+func TestPickBestTag(t *testing.T) {
+	tests := []struct {
+		name      string
+		tag       string
+		manifest  manifest
+		bestTag   string
+		expectErr bool
+	}{
+		{
+			name: "simple lookup",
+			tag:  "v20190329-811f7954b",
+			manifest: manifest{
+				"image1": {
+					TimeCreatedMs: "2000",
+					Tags:          []string{"v20190404-65af07d"},
+				},
+				"image2": {
+					TimeCreatedMs: "1000",
+					Tags:          []string{"v20190329-811f7954b"},
+				},
+			},
+			bestTag: "v20190404-65af07d",
+		},
+		{
+			name: "'latest' overrides date",
+			tag:  "v20190329-811f7954b",
+			manifest: manifest{
+				"image1": {
+					TimeCreatedMs: "2000",
+					Tags:          []string{"v20190404-65af07d"},
+				},
+				"image2": {
+					TimeCreatedMs: "1000",
+					Tags:          []string{"v20190330-811f79999", "latest"},
+				},
+			},
+			bestTag: "v20190330-811f79999",
+		},
+		{
+			name: "tags with suffixes only match other tags with the same suffix",
+			tag:  "v20190329-811f7954b-experimental",
+			manifest: manifest{
+				"image1": {
+					TimeCreatedMs: "2000",
+					Tags:          []string{"v20190404-65af07d"},
+				},
+				"image2": {
+					TimeCreatedMs: "1000",
+					Tags:          []string{"v20190330-811f79999-experimental"},
+				},
+			},
+			bestTag: "v20190330-811f79999-experimental",
+		},
+		{
+			name: "unsuffixed 'latest' has no effect on suffixed tags",
+			tag:  "v20190329-811f7954b-experimental",
+			manifest: manifest{
+				"image1": {
+					TimeCreatedMs: "2000",
+					Tags:          []string{"v20190404-65af07d", "latest"},
+				},
+				"image2": {
+					TimeCreatedMs: "1000",
+					Tags:          []string{"v20190330-811f79999-experimental"},
+				},
+			},
+			bestTag: "v20190330-811f79999-experimental",
+		},
+		{
+			name: "suffixed 'latest' has no effect on unsuffixed tags",
+			tag:  "v20190329-811f7954b",
+			manifest: manifest{
+				"image1": {
+					TimeCreatedMs: "2000",
+					Tags:          []string{"v20190404-65af07d"},
+				},
+				"image2": {
+					TimeCreatedMs: "1000",
+					Tags:          []string{"v20190330-811f79999-experimental", "latest-experimental"},
+				},
+			},
+			bestTag: "v20190404-65af07d",
+		},
+		{
+			name: "'latest' with the correct suffix overrides date",
+			tag:  "v20190329-811f7954b-experimental",
+			manifest: manifest{
+				"image1": {
+					TimeCreatedMs: "2000",
+					Tags:          []string{"v20190404-65af07d-experimental"},
+				},
+				"image2": {
+					TimeCreatedMs: "1000",
+					Tags:          []string{"v20190330-811f79999-experimental", "latest-experimental"},
+				},
+			},
+			bestTag: "v20190330-811f79999-experimental",
```

Comment:
why not set the bestTag as "latest-experimental"? This makes it more descriptive.

When we look at the config file (manifest) one would never know exactly why exactly an older tag is being used.

Reply:
We want to pin to a specific version to avoid new images being pushed resulting in things unexpectedly breaking. `latest-experimental` is a moving target.

Sometimes we really _do_ want that moving target, in which case we would use `latest-experimental` directly (and this tool would not touch such references).